# **CSCE 5218 / CSCE 4930 Deep Learning**

# **The Perceptron** (20 pt)


In [4]:
# Get the datasets
!!curl --output test.dat https://raw.githubusercontent.com/huangyanann/CSCE5218/refs/heads/main/test_small.txt
!!curl --output train.dat https://raw.githubusercontent.com/huangyanann/CSCE5218/refs/heads/main/train.txt


['  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current',
 '                                 Dload  Upload   Total   Spent    Left  Speed',
 '',
 '  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0',
 '100 11645  100 11645    0     0  84958      0 --:--:-- --:--:-- --:--:-- 86902']

In [8]:
# Take a peek at the datasets

# Displaying first 10 lines of train.dat
with open("train.dat", "r") as f:
    for _ in range(10):
        print(f.readline().strip())

A1	A2	A3	A4	A5	A6	A7	A8	A9	A10	A11	A12	A13
1	1	0	0	0	0	0	0	1	1	0	0	1	0
0	0	1	1	0	1	1	0	0	0	0	0	1	0
0	1	0	1	1	0	1	0	1	1	1	0	1	1
0	0	1	0	0	1	0	1	0	1	1	1	1	0
0	1	0	0	0	0	0	1	1	1	1	1	1	0
0	1	1	1	0	0	0	1	0	1	1	0	1	1
0	1	1	0	0	0	1	0	0	0	0	0	1	0
0	0	0	1	1	0	1	1	1	0	0	0	1	0
0	0	0	0	0	0	1	0	1	0	1	0	1	0


In [9]:
# Displaying first 10 lines of test.dat
with open("test.dat", "r") as f:
    for _ in range(10):
        print(f.readline().strip())

X1	X2	X3
1	1	1	1
0	0	1	1
0	1	1	0
0	1	1	0
0	1	1	0
0	1	1	0
0	1	1	0
0	1	1	0
1	1	1	1


### Build the Perceptron Model

You will need to complete some of the function definitions below.  DO NOT import any other libraries to complete this. 

In [10]:
import math
import itertools
import re

In [11]:
# Corpus reader, all columns but the last one are coordinates;
#   the last column is the label
def read_data(file_name):
    f = open(file_name, 'r')

    data = []
    # Discard header line
    f.readline()
    for instance in f.readlines():
        if not re.search('\t', instance): continue
        instance = list(map(int, instance.strip().split('\t')))
        # Add a dummy input so that w0 becomes the bias
        instance = [-1] + instance
        data += [instance]
    return data


In [12]:
def dot_product(array1, array2):
    #TODO: Return dot product of array 1 and array 2

    paired_val = zip(array1, array2)
    products = []

    for a,b in paired_val:
        multiply = a * b
        products.append(multiply)

    result = sum(products)
    
    return result 


In [13]:
def sigmoid(x):
    #TODO: Return outpout of sigmoid function on x
    sig_result = 1 / (1 + math.exp(-x))
    
    return sig_result

# The output of the model, which for the perceptron is 
# the sigmoid function applied to the dot product of 
# the instance and the weights
def output(weight, instance):
    #TODO: return the output of the model 
    dot_prod = dot_product(weight, instance)
    out = sigmoid(dot_prod)
    
    return out


In [14]:
# Predict the label of an instance; this is the definition of the perceptron
# you should output 1 if the output is >= 0.5 else output 0
def predict(weights, instance):
    #TODO: return the prediction of the model
    dot_prod = dot_product(weights, instance)
    model_out = sigmoid(dot_prod)

    if model_out >= 0.5:
        return 1
    else:
        return 0


# Accuracy = percent of correct predictions
def get_accuracy(weights, instances):
    # You do not to write code like this, but get used to it
    correct = sum([1 if predict(weights, instance) == instance[-1] else 0
                   for instance in instances])
    return correct * 100 / len(instances)


In [15]:
# Train a perceptron with instances and hyperparameters:
#       lr (learning rate) 
#       epochs
# The implementation comes from the definition of the perceptron
#
# Training consists on fitting the parameters which are the weights
# that's the only thing training is responsible to fit
# (recall that w0 is the bias, and w1..wn are the weights for each coordinate)
#
# Hyperparameters (lr and epochs) are given to the training algorithm
# We are updating weights in the opposite direction of the gradient of the error,
# so with a "decent" lr we are guaranteed to reduce the error after each iteration.
def train_perceptron(instances, lr, epochs):

    #TODO: name this step
    # Answer - Initializing the weights
    weights = [0] * (len(instances[0])-1)

    for _ in range(epochs):
        for instance in instances:
            #TODO: name these steps
            # Answer - Forward Pass
            in_value = dot_product(weights, instance)
            output = sigmoid(in_value)
            error = instance[-1] - output
            #TODO: name these steps
            # Answer - Backpropagation
            for i in range(0, len(weights)):
                weights[i] += lr * error * output * (1-output) * instance[i]

    return weights

## Run it

In [16]:
instances_tr = read_data("train.dat")
instances_te = read_data("test.dat")
lr = 0.005
epochs = 5
weights = train_perceptron(instances_tr, lr, epochs)
accuracy = get_accuracy(weights, instances_te)
print(f"#tr: {len(instances_tr):3}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
      f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 400, epochs:   5, learning rate: 0.005; Accuracy (test, 14 instances): 71.4


## Questions

Answer the following questions. Include your implementation and the output for each question.



### Question 1

In `train_perceptron(instances, lr, epochs)`, we have the follosing code:
```
in_value = dot_product(weights, instance)
output = sigmoid(in_value)
error = instance[-1] - output
```

Why don't we have the following code snippet instead?
```
output = predict(weights, instance)
error = instance[-1] - output
```

#### TODO Add your answer here (text only)




### Answer
#### For the predict() function, the return value is either 0 or 1. But the sigmoid() function returns a floating value within continuous range of 0 to 1. In our case, we need the return value to be a continuous value within the range 0 to 1 which is needed for gradient descent to calculate the error and update the weights accordingly. So we need to use the sigmoid() function.
#### Moreover the predict() function is mainly used to give a prediction value, whereas the sigmoid() function is used as an activation function here for the training phase.

### Question 2
Train the perceptron with the following hyperparameters and calculate the accuracy with the test dataset.

```
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]              # number of epochs
lr = [0.005, 0.01, 0.05]              # learning rate
```

TODO: Write your code below and include the output at the end of each training loop (NOT AFTER EACH EPOCH)
of your code.The output should look like the following:
```
# tr:  20, epochs:   5, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  10, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
# tr:  20, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
[and so on for all the combinations]
```
You will get different results with different hyperparameters.

#### TODO Add your answer here (code and output in the format above) 


In [17]:
instances_tr = read_data("train.dat")
instances_te = read_data("test.dat")
tr_percent = [5, 10, 25, 50, 75, 100] # percent of the training dataset to train with
num_epochs = [5, 10, 20, 50, 100]     # number of epochs
lr_array = [0.005, 0.01, 0.05]        # learning rate

for lr in lr_array:
  for tr_size in tr_percent:
    for epochs in num_epochs:
      size =  round(len(instances_tr)*tr_size/100)
      pre_instances = instances_tr[0:size]
      weights = train_perceptron(pre_instances, lr, epochs)
      accuracy = get_accuracy(weights, instances_te)
      print(f"#tr: {len(pre_instances):0}, epochs: {epochs:3}, learning rate: {lr:.3f}; "
            f"Accuracy (test, {len(instances_te)} instances): {accuracy:.1f}")

#tr: 20, epochs:   5, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 20, epochs:  10, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 20, epochs:  20, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 20, epochs:  50, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 20, epochs: 100, learning rate: 0.005; Accuracy (test, 14 instances): 85.7
#tr: 40, epochs:   5, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 40, epochs:  10, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 40, epochs:  20, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 40, epochs:  50, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 40, epochs: 100, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 100, epochs:   5, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 100, epochs:  10, learning rate: 0.005; Accuracy (test, 14 instances): 71.4
#tr: 100, epochs:  20, learning rate: 

### Question 3
Write a couple paragraphs interpreting the results with all the combinations of hyperparameters. Drawing a plot will probably help you make a point. In particular, answer the following:
- A. Do you need to train with all the training dataset to get the highest accuracy with the test dataset?
- B. How do you justify that training the second run obtains worse accuracy than the first one (despite the second one uses more training data)?
   ```
#tr: 100, epochs:  20, learning rate: 0.050; Accuracy (test, 100 instances): 71.0
#tr: 200, epochs:  20, learning rate: 0.005; Accuracy (test, 100 instances): 68.0
```
- C. Can you get higher accuracy with additional hyperparameters (higher than `80.0`)?
- D. Is it always worth training for more epochs (while keeping all other hyperparameters fixed)?

#### TODO: Add your answer here (code and text)



#### The conclusions that we can see from the output given in Question 2 are :
 - Inference from Training Instances :
   > For some cases, using 200 or more tr with lr of 0.005 and 0.05 gave the highest accuracy of 85.7%.
   > 
   > For some cases, increasing the tr without adjusting other hyperparameters still resulted in the accuracy of 71.4% or less and did not increase.
 - Inference from Epochs :
   > In most of the cases, increasing the number of epoch does not really increase the accuracy more than 71.4%.
   > 
   > Only in very few cases having higher epochs like 100 gives the highest accuracy of 85.7% when the learning rate is 0.005 and 200 training insrances are taken with it.
   > 
   > In some cases, when the epoch is increased with learning rate of 0.01 and 0.05 and training instances of 100, it is reducing the accuracy very much to 28.6%.
 - Inference from Learning Rate :
   > When the learning rate is low (0.005) it takes too many epochs to get accuracy max up to 71.4% and sometimes upto 85.7%.
   > 
   > When learning rate is high, the training becomes unstable and in multiple cases, it drops the accuracy to even 28.6%
   > 
   > When the learning is moderate, it is able to reach the highest accuracy of 85.7% but again drops the accuracy when the epochs are set too high for this learning rate.


#### Answer A.
No. We do not need to train with the whole dataset to reach the highest accuracy.

> One such example : #tr: 20, epochs:  50, learning rate: 0.010; Accuracy (test, 14 instances): 85.7

Here we are only using 20 training instances and getting the highest accuracy.

This shows that a well chosen subset of the training datasert can perform as good as the whole dataset or even better than that.

#### Answer B.
At any point of running the code if the accuracy drops low when it uses more training data as compared to a scenario where it gives better accuracy with lesser training data, it is mainly because of the overfitting. It might also be an inference that most of the data is not informative or noisy.

#### Answer C.
Yes. The accuracy can be increased in the following ways.
> Firstly, like we can use Adam or RMSProp optimizer instead of SGD.
> 
> Secondly, we can use a L2 Regularization to prevent overly large weights.
>
> Thirdly, we can use dropout to randomly drop neurons to get better generalization of the data.

#### Answer D.
No. It is not always necessary to train for more epochs. Sometimes, higher number of epochs can even reduce the accuracy due to overfitting.
So we use a mechanism to stop the epochs at an early stage than what we set initially, when the error does not change much. This mechanism is called early stopping and we set a patience which is the number of epochs still to run even if the error is not changing much. This is mainly used to overcome a saddle point or a local minima.